lib/rust/profiling/src/lib.rs

```diff
@@ -0,0 +1,427 @@
+/// metadata.
+pub fn mark_start_interval(metadata: Metadata) -> IntervalHandle {
+    let interval_name = start_interval_label(&metadata);
+    if profiling_level_is_active(metadata.profiling_level.clone()) {
+        mark_with_metadata(interval_name.into(), metadata.clone().into());
+    }
+    IntervalHandle::new(metadata)
+}
+
+fn get_latest_performance_entry() -> Option<PerformanceEntry> {
+    let entries: js_sys::Array = performance().get_entries_by_type("measure");
+
+    if entries.length() < 1 {
+        return None;
+    }
+    let measure = entries.get(entries.length() - 1);
+    let measure: PerformanceEntry = measure.into();
+    Some(measure)
+}
+
+/// Mark the end of an measuring an interval in the JS API. Return the measurement taken between
+/// start and end of the interval, if possible.
+pub fn mark_end_interval(metadata: Metadata) -> Result<Measurement, MeasurementError> {
+    let profiling_level = metadata.profiling_level.clone();
+    let start_label = start_interval_label(&metadata);
+    let end_label = end_interval_label(&metadata);
+    let measurement_label = measure_interval_label(&metadata);
+    if !profiling_level_is_active(profiling_level) {
+        Err(MeasurementError::ProfilingDisabled)
+    } else {
+        mark_with_metadata(end_label.clone().into(), metadata.clone().into());
+        measure_with_start_mark_and_end_mark_and_metadata(
+            measurement_label.into(),
+            start_label.into(),
+            end_label.into(),
+            metadata.into(),
+        );
+
+        let measure = get_latest_performance_entry().ok_or(MeasurementError::NoMeasurementFound)?;
+
+        let measurement: Measurement =
+            measure.try_into().or(Err(MeasurementError::InvalidMeasurementConversion))?;
+        Ok(measurement)
+    }
+}
+
+
+
+// ======================
+// === IntervalHandle ===
+// ======================
+
+/// Handle that allows ending the interval.
+#[derive(Clone, Debug)]
+pub struct IntervalHandle {
+    metadata: Metadata,
+    released: bool,
+}
+
+impl IntervalHandle {
+    fn new(metadata: Metadata) -> Self {
+        IntervalHandle { metadata, released: false }
+    }
+
+    /// Measure the interval.
+    pub fn end(mut self) {
+        self.released = true;
+        warn_on_error(mark_end_interval(self.metadata.clone()));
+    }
+
+    /// Release the handle to prevent a warning to be emitted when it is garbage collected without
+    /// a call to `end`. This can be useful if one wants to call `end_interval` manually, or the
+    /// equivalent call to `end_interval` is in Rust code.
+    pub fn release(mut self) {
+        self.released = true;
+        drop(self)
+    }
+}
+
+impl Drop for IntervalHandle {
+    fn drop(&mut self) {
+        if !self.released {
+            warn_on_error(mark_end_interval(self.metadata.clone()));
+            WARNING!(format!("{} was dropped without a call to `measure`.", self.metadata.label));
```

Comment:
The warning message mentions a `measure` method, but there is no such method in the code.